lotus-soup/testkit/role_miner_sync.go

```diff
@@ -0,0 +1,185 @@
+package testkit
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/testground/sdk-go/run"
+	"github.com/testground/sdk-go/sync"
+)
+
+// MinerSyncTopic is the topic where miner events are exchanged.
+var MinerSyncTopic = sync.NewTopic("miner_sync", &MiningSyncMsg{})
+
+// MiningSyncMsg signals that the miner attempted to mine in a particular epoch,
+// indicating whether it produced a block or not.
+type MiningSyncMsg struct {
+	// Addr is the address of the miner.
+	Addr address.Address
+	// Produced signals whether the miner produced a block or not.
+	Produced bool
+}
+
+// SynchronizeMiner synchronizes this miner with all other miners in this test,
+// so that they're mining in perfect synchrony according to a globally-
+// synchronized epoch clock.
+//
+// This function _receives_ "epoch start" signals and _emits_ "epoch end"
+// signals:
+//
+//  * an epoch start signal attempts to locally mine a block, and publishes
+//    a message on a sync topic indicating that we're done mining, and whether
+//    we produced a block or not.
+//  * all miners wait until their HEAD advances to the tipset that includes
+//    all blocks that were produced in this epoch.
+//  * once that happens, we emit an epoch end signal, which the caller can
+//    interpret as a signal to advance the clock and begin a new epoch.
+func (m *LotusMiner) SynchronizeMiner(ctx context.Context, epochStart <-chan abi.ChainEpoch, epochEnd chan<- abi.ChainEpoch) {
```

Comment:
`epochStart`:
* what do we start? Mining. It seems basically a "go mine" signal.
* epoch is confusing, it may seem to imply that we mine on the epoch passed in the channel when in fact the epoch is used just for debugging purposes
* the channel seems to be doing to things at once: signal to mine (this could just be an empty struct) and _also_ tell you in which epoch the miner is standing on (do we need to obtain that information from here?)

Reply:
`epochEnd`: similar, do we even read the epoch we're sending? or is this just i'm done syncing, the terms epoch and end seem to imply to me _we are controlling the clock here_ but i think this is just a signal that i (miner) did something (either mined or synced the chain, unrelated to start/end epoch)

Reply:
Basically this routine is intended to behave like an "actor", as in the actor-oriented programming model programming. It receives signals to work on `epochStart` and emits "work done" signals on `epochEnd`. In Akka, you would assimilate the `epochStart` channel to an inbox, and `epochEnd` to an outbox (using their mailbox analogy).

I haven't thought too much about how these epochs are actually used; I started making some of these channels of element type `struct{}`, but then I realised that I might as well propagate useful info. Although in hindsight, I see how that decision (and loose usage of the sent/recvd values) might confuse others.

How about we name these channels `mineStart` and `mineDone`? From the viewpoint of this routine, it establishes the clear boundary of signals that demarcate each unit of work.

Reply:
Those names sound good.